File: hw3.ml
(* CSE305 Spring 2014
 * A possible solution to HW2, expressed in Standard ML
 * Author: Carl Alphonce
 * Updated by: Sean Frischmann
 *)

(****************************************************************************
 TYPE DEFINITIONS 
 ****************************************************************************)

(* The type of expressions and special forms *)
datatype exp =
    Boolean of bool
  | Number of int
  | String of string
  | Name of string
  | Error | Quit | Add | Sub | Mul | Div | Rem | Pop | Exc | Neg
  | And | Or | Not | Equal | LessThan | If | Bind;

(****************************************************************************
 UTILITY FUNCTIONS 
 ****************************************************************************)

(* Utility function to build a textual form of an expression
   or special form, used for printing.
 *)
fun expression2string(Boolean(true)) = ":true:"
  | expression2string(Boolean(false)) = ":false:"
  | expression2string(Number(X)) = if (X<0) then "-"^Int.toString(~X) else Int.toString(X)
  | expression2string(Error) = ":error:"
  | expression2string(String(X)) = "\""^X^"\""
  | expression2string(Name(X)) = X
  | expression2string(_) = "?? should not happen ??"
  ;

(* Utility function to print stack *)
fun printStack([]) = ()
  | printStack(x::xs) = ( TextIO.print(expression2string(x)^"\n") ; printStack(xs) );

(* Computes quotient according to rules of the language *)
fun quotient(a2,a1) =
    if (a1<0)
    then if (a2 div a1) * a1 > a2 then (a2 div a1)+1 else (a2 div a1)
    else (a2 div a1);

(* Computes remainder according to rules of the language *)
fun remainder(a2,a1) = a2 - a1 * quotient(a2,a1);

(* Computes primitive boolean functions*)
fun booleanFunctions(And, a2, a1) = if (a2 = a1) then true else false
   | booleanFunctions(Or, a2, a1) = if (a2 = a1) then false else true
   | booleanFunctions(Not, a2, a1) = if (a1 = true) then false else true;

(* Computes primitive relational operators*)
fun relationalOperators(Equal, a2, a1) = if (a2 = a1) then true else false
   | relationalOperators(LessThan, a2, a1) = if (a2 < a1) then true else false

(* Computes primitive if function*)
fun ifFunction(a1,a2,a3) = 
    if (a1 = true) then a3
    else a2;

(****************************************************************************
 MAIN EVALUATOR FUNCTIONS 
 ****************************************************************************)

(* Applies a primitive function to its two arguments *)

fun 
  (* 0-ARY OPERATORS *)
    apply(_,[])  = (Error,[])

  (* UNARY OPERATORS *)
  | apply(Neg,Number(a)::rest) = (Number(~a),rest)
  | apply(Not,Boolean(a1)::rest) = (Boolean(booleanFunctions(Not,true,a1)),rest)
  | apply(_,[x]) = (Error, [x])

  (* BINARY OPERATORS *)
  | apply(Add,Number(a1)::Number(a2)::rest) = (Number(a2+a1),rest)
  | apply(Sub,Number(a1)::Number(a2)::rest) = (Number(a2-a1),rest)
  | apply(Mul,Number(a1)::Number(a2)::rest) = (Number(a2*a1), rest)
  | apply(Div,Number(a1)::Number(a2)::rest) = (Number(quotient(a2,a1)),rest)
  | apply(Rem,Number(a1)::Number(a2)::rest) = (Number(remainder(a2,a1)),rest)
  | apply(And,Boolean(a1)::Boolean(a2)::rest) = (Boolean(booleanFunctions(And,a2,a1)),rest)
  | apply(Or,Boolean(a1)::Boolean(a2)::rest) = (Boolean(booleanFunctions(Or,a2,a1)),rest)
  | apply(Equal,Number(a1)::Number(a2)::rest) = (Boolean(relationalOperators(Equal,a2,a1)),rest)
  | apply(LessThan,Number(a1)::Number(a2)::rest) = (Boolean(relationalOperators(LessThan,a2,a1)),rest)
  | apply(If,Boolean(a1)::a2::a3::rest) = (ifFunction(a1,a2,a3),rest)
  | apply(_,stack) = (Error,stack)
  ;

(* stack operations *)

fun stackOps(Pop, x::stack) = stack
  | stackOps(Exc, x::y::stack) = y::x::stack
  | stackOps(_,stack) = Error::stack
  ;

(* Help Evaluates an expression *)
fun evalHelper(Boolean(x),stack) = Boolean(x)::stack
  | evalHelper(Number(x),stack) = Number(x)::stack
  | evalHelper(String(x),stack) = String(x)::stack
  | evalHelper(Name(x),stack) = Name(x)::stack
  | evalHelper(Quit,stack) = Quit::stack
  | evalHelper(Pop,x::stack) = stack
  | evalHelper(Exc,x::y::stack) = y::x::stack
  | evalHelper(Error,stack) = Error::stack
  | evalHelper(expr, stack) = let
            val (v,s) = apply(expr, stack)
      in
            v::s
      end;

(*Helps sort out stack after bind*)
fun bindStackHelper(a1, stack) = a1::stack;

(*Helps sort out bindings*)
fun bindBindingsHelper(a2, bindings) = a2::bindings;

fun bindValuesHelper(a1, values) = a1::values;

fun existCheck(a2, []) = false
  | existCheck(Name(a2), Name(x)::bindings) = if (Name(a2) = Name(x)) then true else existCheck(Name(a2), bindings)
  | existCheck(a2, x::bindings) = if (a2 = x) then true else existCheck(a2, bindings)
;

fun nameCheck(a1::Name(a2)::stack) = true
  | nameCheck(_) = false
;

fun checkIfName(Name(a1)) = true;

fun findValue(Name(a2), Name(x)::bindings, y::values) = 
    if (Name(a2) = Name(x)) then y
    else findValue(Name(a2), bindings, values)
;

fun bindFailure(a1::Name(a2)::stack) = a1::stack
  | bindFailure(Name(a1)::stack) = stack
  | bindFailure(stack) = stack
;

(* Evaluates Bind*)
fun bindFunction(a1::a2::stack, bindings::values::rest) = 
     if existCheck(a2, bindings)
         then ((evalHelper(Error, stack))::bindings::values::rest)
     else ((bindStackHelper(a1,stack))::(bindBindingsHelper(a2,bindings)::bindValuesHelper(a1, values)::rest))
  | bindFunction(a1::stack,bindings::rest) = ((evalHelper(Error, stack))::bindings::rest)
  | bindFunction(stack,bindings::rest) = ((evalHelper(Error, stack))::bindings::rest)
;

(* Evaluates an expression *)

fun eval(Boolean(x), environment) = ((evalHelper(Boolean(x),hd environment))::(tl environment))
  | eval(Number(x), environment)  = ((evalHelper(Number(x),hd environment))::(tl environment))
  | eval(String(x), environment)  = ((evalHelper(String(x),hd environment))::(tl environment))
  | eval(Name(x), stack::bindings::values::rest)  = 
      if existCheck(Name(x),bindings) then (
          (findValue(Name(x),bindings,values)::stack)::(bindings::values::rest))
      else ((evalHelper(Name(x), stack))::(bindings::values::rest))
  | eval(Quit, environment)       = ((evalHelper(Quit,hd environment))::(tl environment))
  | eval(Pop, environment)     = ((evalHelper(Pop,hd environment))::(tl environment))
  | eval(Exc, environment )     = ((evalHelper(Exc,hd environment))::(tl environment))
  | eval(Error, environment)      = ((evalHelper(Error,hd environment))::(tl environment))
  | eval(Bind, stack::bindings)      = 
      if nameCheck(stack) then (bindFunction(stack, bindings))
      else ((evalHelper(Error, bindFailure(stack)))::(bindings))
  | eval(expr, environment ) = ((evalHelper(expr,hd environment))::(tl environment))
  ;

(* Functions to parse a number *)

fun parseNumber(x,inStr) = 
    case (TextIO.input1(inStr)) of
	NONE => SOME(Error)
      | SOME(ch) => 
	    if (Char.isDigit(ch)) then parseNumber(x*10+(ord(ch)-ord(#"0")),inStr)
       else if (Char.isSpace(ch)) then SOME(Number(x))
       else SOME(Error)
      ;

fun parseNegativeNumber(inStr) = 
    case ( parseNumber(0,inStr) ) of
        NONE => SOME(Error)
      | SOME(Number(num)) =>  SOME(Number(~1 * num))
      | SOME(_) => SOME(Error) 
      ;

(* Function to parse a boolean  *)

fun parseBoolean(x, inStr) = 
    case (TextIO.input1(inStr)) of
	NONE => SOME(Error)
      | SOME(ch) => 
	    if (Char.isAlpha(ch) orelse ch = #":") then parseBoolean(x^Char.toString(ch),inStr)
       else if (Char.isSpace(ch))
	       then if (x = ":true:")  then SOME(Boolean(true))
	       else if (x = ":false:") then SOME(Boolean(false))
	       else SOME(Error)
       else SOME(Error);

(* Function to parse a string  *)

fun parseString(x, inStr, y) = 
    case (TextIO.input1(inStr)) of
	NONE => SOME(Error)
      | SOME(ch) => 
       if (ch = #"\"") then parseString(x, inStr, (y+1))
       else if (Char.isAlpha(ch)) then parseString(x^Char.toString(ch), inStr, y)
       else if (Char.isSpace(ch)) then
            if (y < 1) then parseString(x^Char.toString(ch), inStr, y)
            else SOME(String(x))
       else SOME(Error);

(* Function to parse an error *)

fun parseError(x, inStr) = 
    case (TextIO.input1(inStr)) of
	NONE => SOME(Error)
      | SOME(ch) => 
	    if (Char.isAlpha(ch) orelse ch = #":") then parseError(x^Char.toString(ch),inStr)
       else if (Char.isSpace(ch))
	       then if (x = ":error:")  then SOME(Error)
	       else SOME(Error)
       else SOME(Error);

(***** EDITED TO HERE ********************************************************************)

(* Function to parse a boolean or error  *)

fun parseBooleanOrError(x, inStr) = 
    case (TextIO.input1(inStr)) of
	NONE => SOME(Error)
      | SOME(ch) => 
	    if (ch = #"e")                  then parseError(x^Char.toString(ch),inStr)
       else if (ch = #"t" orelse ch = #"f") then parseBoolean(x^Char.toString(ch),inStr)
       else SOME(Error);

(* Function to parse a primitive  *)

fun parsePrimitive(x, inStr) = 
    case (TextIO.input1(inStr)) of
	NONE => SOME(Error)
      | SOME(ch) => 
	    if (Char.isAlpha(ch))  then parsePrimitive(x^Char.toString(ch),inStr)
       else if (Char.isSpace(ch))
	       then if (x = "add") then SOME(Add)
	       else if (x = "sub") then SOME(Sub)
	       else if (x = "mul") then SOME(Mul)
	       else if (x = "div") then SOME(Div)
	       else if (x = "rem") then SOME(Rem)
	       else if (x = "pop") then SOME(Pop)
	       else if (x = "exc") then SOME(Exc)
	       else if (x = "neg") then SOME(Neg)
	       else if (x = "and") then SOME(And)
	       else if (x = "or") then SOME(Or)
	       else if (x = "not") then SOME(Not)
	       else if (x = "equal") then SOME(Equal)
	       else if (x = "lessThan") then SOME(LessThan)
	       else if (x = "if") then SOME(If)
	       else if (x = "bind") then SOME(Bind)
               else if (x = "quit") then SOME(Quit)
	       else SOME(Name(x))
       else SOME(Error);

(* A recursive helper function for the parse function, which reads
   a character from the input stream and determines what more
   specific parsing function to call.
 *)      
fun parseHelper(NONE, inStr) = NONE
  | parseHelper(SOME(ch), inStr) =
      if (Char.isDigit(ch)) then parseNumber(ord(ch)-ord(#"0"),inStr)
 else if (ch = #"-")        then parseNegativeNumber(inStr)
 else if (ch = #":")        then parseBooleanOrError(":", inStr)
 else if (ch = #"\"")        then parseString("", inStr, 0)
 else if (Char.isAlpha(ch)) then parsePrimitive(Char.toString(ch), inStr)
 else if (Char.isSpace(ch)) then parseHelper(TextIO.input1(inStr), inStr)
 else NONE;

fun maybePrintPrompt(stack::bindings::values::rest, SOME(0)) = ( printStack(stack); TextIO.print("repl> ") )
  | maybePrintPrompt(stack, SOME(_)) = ();

(* Function to parse the next expression on the input stream. *)      
fun parse(inStr) = parseHelper(TextIO.input1(inStr), inStr);

fun replHelper(inStr, environment) =
(
    maybePrintPrompt(environment, TextIO.canInput(inStr,1));
    case (parse(inStr)) of 
       NONE => replHelper(inStr, environment)
     | SOME(Quit) => () 
     | SOME(expression) => replHelper(inStr,eval(expression, environment)) 
);


fun repl() = replHelper(TextIO.stdIn, [[],[],[]]);

File: hw3.c
/* CSE305 Spring 2014
 * A possible solution to HW2, expressed in C
 * Author: Carl Alphonce & Daniel Bellinger 
 * 
 * New Features Added by: Sean Frischmann
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include "Node.c"
#include "Type.c" 

#define BUFFER_LENGTH 256


/*
 * Data for bindings
 * New Feature
 */
typedef struct Data{
	 char *key;
	 struct Type * value;
	 struct Data *next;
}data_points;

/*
 * Map for bindings
 * New Feature
 */
typedef struct Map{
	 struct Data * start;
}map;

/*
 * Add bindings to Map
 * New Feature
 */
void addToMap(map *list, char *key, struct Type *value){
	data_points *temp;
	temp = (struct Data *) (malloc(sizeof(struct Data)));
	temp->key = key;
	temp->value = value;
	temp->next = list->start;
	list->start = temp;
}

/*
 * Find Key
 * New Feature
 */
int findInMap(map *list, char *key){
	 data_points *current = list->start;
	 while(current != NULL){
		if(strcmp(current->key, key)==0) {return TRUE;}
			current = current->next;
	 }
	 return FALSE;
}

struct Type * makeError() {
	 struct Type * x = (struct Type *) (malloc(sizeof(struct Type)));;
	 (*x).type = ERROR;
	 (*x).value.error = ERRORKIND;
	 return x;
}

/*
 * Find Value
 * New Feature
 */
struct Type * findValue(map *list, char *key){
	 data_points * current = list->start;
	 struct Type * temp;
	 while (current != NULL){
		  if(strcmp(current->key, key)==0) { 
				if((current->value->type == NAME)&&(findInMap(list, current->value->value.name))){
					key = current->value->value.name;
					current = list->start;
					continue;
				}else{
					 temp = current->value;
				}return temp;
		  }
		  current = current->next;
	 }
	 return makeError();
}

struct Type * makeNumber(int num) {
    struct Type * px = (struct Type *) (malloc(sizeof(struct Type)));;
    (*px).type = NUMBER;
    (*px).value.number = num;
    return px;
}

struct Type * makeBoolean(int b) {
    struct Type * px = (struct Type *) (malloc(sizeof(struct Type)));;
    (*px).type = BOOLEAN;
    (*px).value.bool = b;
    return px;
}
/*
 * Creates Name
 *   if name is binded uses value instead
 * New Feature
 */
struct Type * makeName(char *b, map *bindings) {
	 if(findInMap(bindings, b) == TRUE){
		 return findValue(bindings, b);
	 }
	 struct Type * px;
	 px = (struct Type *) (malloc(sizeof(struct Type)));;
	 char * temp = (char *) (malloc(strlen(b)+1));
	 strncpy(temp,b, strlen(b)+1);
	 (*px).type = NAME;
	 (*px).value.string = temp;
	 return px;
}

/*
 * Creates string
 * New Feature
 */
struct Type * makeString(char * b) {
	 struct Type * px = (struct Type *) (malloc(sizeof(struct Type)));;
	 char * temp = (char *) (malloc(strlen(b)));
	 int i;
	 int j = 0;
	 int length = strlen(b)-1;
	 for(i=1;i<length;i++){
		  temp[j] = b[i];
		  j = j+1;
	 }
	 (*px).type = STRING;
	 (*px).value.string = temp;
	 return px;
}

int isBoolean(struct Type * t) { return (*t).type == BOOLEAN; }
int isError(struct Type * t)   { return (*t).type == ERROR; }
int isNumber(struct Type * t)  { return (*t).type == NUMBER; }
int valueOf(struct Type * t)   { return (*t).value.number; }
/*
 * Checks if type is name
 * New Feature
 */
int isName(struct Type * t) { return (*t).type == NAME; }
/*
 * Checks if type is string
 * New Feature
 */
int isString(struct Type * t) { return (*t).type == STRING; }

/*
 * Checks if token is a letter
 * New Feature
 */
int isLetter(char * token) { 
	 if(((token[0]-0)>64)&&((token[0]-0)<91)){return TRUE;}
	 else if(((token[0]-0)>96)&&((token[0]-0)<123)) {return TRUE;}
	 else{return FALSE;}
}

/*   Pop top value from stack	*/
void popStack(struct Node *stack){
  struct Type * t;
  if((t = pop(stack)) == NULL){
    push(stack,makeError());
  }
}

/*   Swap top 2 values of stack	  */   

void excStack(struct Node *stack){
  struct Type * x = pop(stack);
  struct Type * y = pop(stack);
  //make sure there are 2 values to exchange
  if(x == NULL){ //no values on the stack
    push(stack,makeError());
  }else if(y == NULL){ //only 1 value was on the stack
  	push(stack,x);
  	push(stack,makeError());
  }else{
  	push(stack, x);
  	push(stack, y);
  }
}

/*   negate value   */

struct Type * negate(struct Type * number){
  if (isNumber(number)) { return makeNumber(-valueOf(number)); }
  return makeError();
}

/* Bind function
 * New Feature
 */ 

struct Type * bind(struct Type * x, struct Type * y){
  if (isName(x)) {

  }
  return makeError();
}

/* And function
 * New Feature
 */ 

struct Type * and(struct Type * x, struct Type * y){
  if (isBoolean(x) && isBoolean(y)) {
		if ((*x).value.bool == (*y).value.bool) { return makeBoolean(TRUE);}
		else {return makeBoolean(FALSE);}
  }
  return makeError();
}

/* or function
 * New Feature
 */ 

struct Type * or(struct Type * x, struct Type * y){
  if (isBoolean(x) && isBoolean(y)) {
		if ((*x).value.bool != (*y).value.bool) { return makeBoolean(TRUE);}
		else {return makeBoolean(FALSE);}
  }
  return makeError();
}

/* not function
 * New Feature
 */ 

struct Type * not(struct Type * x){
  if (isBoolean(x)) {
		if ((*x).value.bool == 0) { return makeBoolean(TRUE);}
		else {return makeBoolean(FALSE);}
  }
  return makeError();
}

/* equal function
 * New Feature
 */ 

struct Type * equal(struct Type * x, struct Type * y){
  if (isNumber(x) && isNumber(y)) {
		if ((*x).value.bool == (*y).value.bool) { return makeBoolean(TRUE);}
		else {return makeBoolean(FALSE);}
  }
  return makeError();
}

/* lessThan function
 * New Feature
 */ 

struct Type * lessThan(struct Type * x, struct Type * y){
  if (isNumber(x) && isNumber(y)) {
		if ((*x).value.bool > (*y).value.bool) { return makeBoolean(TRUE);}
		else {return makeBoolean(FALSE);}
  }
  return makeError();
}

/*   Add function   */

struct Type * add(struct Type * num1, struct Type * num2){
  if (isNumber(num1) && isNumber(num2)) {
    return makeNumber(valueOf(num1) + valueOf(num2));
  }
  return makeError();
}

/*   Subtract function	 */

struct Type * sub(struct Type * num1, struct Type * num2){
  if (isNumber(num1) && isNumber(num2)) {
    return makeNumber(valueOf(num1) - valueOf(num2));
  }
  return makeError();
}


/*  Multiply function	*/

struct Type * mul(struct Type * num1, struct Type * num2){
  if (isNumber(num1) && isNumber(num2)) {
    return makeNumber(valueOf(num1) * valueOf(num2));
  }
  return makeError();
}

/*   Divide function   */

int c_divide(int a, int b) {
  if (b<0) { return b/a + (a<0? 1 : -1); }
  else     { return b/a; }
}

struct Type * divide(struct Type * num1, struct Type * num2){
  if (isNumber(num1) && isNumber(num2)) {
    return makeNumber( c_divide(valueOf(num1), valueOf(num2)) );
  }
  return makeError();
}

/*   Remainder function	  */

struct Type * rem(struct Type * num1, struct Type * num2){
  if (isNumber(num1) && isNumber(num2)) {
    return makeNumber( valueOf(num2) - valueOf(num1) * valueOf(divide(num1,num2)) );
  }
  return makeError();
}


/* toString() for each type   */

char * expression2string(struct Type * t){
  char * str = (char *) malloc(sizeof(char)*255);
  switch( (*t).type ){
  case NUMBER:
    sprintf(str,"%d",(*t).value.number);
    break;
  case BOOLEAN:
    str = (*t).value.bool == TRUE ? ":true:" : ":false:";
    break;
  case ERROR:
    str = ":error:";
    break;
  case STRING:
    sprintf(str,"\"%s\"",(*t).value.string);
    break;
  case NAME:
    sprintf(str,"%s",(*t).value.string);
    break;
  }   
  return str;
}


/*   Print out the stack   */

void printStack(struct Node *stack){
  // loop through stack list, toString each value, print
  stack = stack->cdr; // skip past header node
  while(stack != NULL){
    printf("%s\n",expression2string(stack->car));
    stack = stack->cdr;
  }
}

/*   Apply different types of operators   */

void applyBinary(struct Type * (*f)(), struct Node * stack) {
  struct Type * x = pop(stack);
  struct Type * y = pop(stack);
  if (y != NULL && x != NULL) {
    struct Type * result = (*f)(x,y);
    if ((*result).type == ERROR) {
      push(stack,y);
      push(stack,x);
    }
    push(stack,result);
  }
  else if (y == NULL && x == NULL) {
    // stack was empty
    push(stack,makeError());
  }
  else { 
    // stack had ONE item on it
    push(stack,x);
    push(stack,makeError());
  }
}

void applyIf(struct Node * stack) {
	struct Type * last = pop(stack);
	struct Type * middle = pop(stack);
	struct Type * first = pop(stack);
	if (last != NULL && middle != NULL && first != NULL) {
		if (last->type != BOOLEAN) {
			push(stack,first);
			push(stack,middle);
			push(stack,makeError());
		}else if (last->value.bool == TRUE){
			push(stack,first);
		}else{
			push(stack,middle);
		}
	}
	else if (last == NULL && middle == NULL && first == NULL) {
		// stack was empty
		push(stack,makeError());
	}
	else if (middle == NULL && first == NULL){ 
		// stack had ONE item on it
		push(stack,last);
		push(stack,makeError());
	}else{
		push(stack,middle);
		push(stack,last);
		push(stack,makeError());
	}

}

/*
 * Binds to values from the stack
 * New Feature
 */
void applyBindings(struct Node * stack, map *bindings) {
	 struct Type * x = pop(stack);
	 struct Type * y = pop(stack);
	 if (y != NULL && x != NULL) {
		struct Type * result;
		if ((*y).type == NAME){
			if(findInMap(bindings, y->value.name) == TRUE){
					 result = makeError();
			}else{
				addToMap(bindings, y->value.name, x);
				result = x;
			}
		  }else{
			result = makeError();
	 }
	 if ((*result).type == ERROR) {
		push(stack,y);
		push(stack,x);
	 }
	 push(stack,result);
	 }
	 else if (y == NULL && x == NULL) {
	 // stack was empty
	 push(stack,makeError());
	 }
	 else { 
	 // stack had ONE item on it
	 push(stack,x);
	 push(stack,makeError());
	 }
}

void applyUnary(struct Type * (*f)(), struct Node * stack) {
  struct Type * x = pop(stack);
  if (x != NULL) {
    struct Type * result = (*f)(x);
    if ((*result).type == ERROR) {
      push(stack,x);
    }
    push(stack,result);
  }
  else { // stack was empty
    push(stack,makeError());
  }
}

/*
 * Updated to incorporate new features
 */
struct Node * eval(char * token,struct Node *stack, map *bindings){
	 if(strcmp(token,":true:")  == 0) { push(stack,makeBoolean(TRUE)); }
	 else if(token[0] == '"') { push(stack,makeString(token)); }
	 else if(strcmp(token,":false:") == 0) { push(stack,makeBoolean(FALSE)); }
	 else if(strcmp(token,":error:") == 0) { push(stack,makeError()); }
	 else if(strcmp(token,"equal")== 0)      { applyBinary(equal, stack); }
	 else if(strcmp(token,"lessThan")== 0)      { applyBinary(lessThan, stack); }
	 else if(strcmp(token,"and")== 0)      { applyBinary(and, stack); }
	 else if(strcmp(token,"or")== 0)      { applyBinary(or, stack); }
	 else if(strcmp(token,"not")== 0)      { applyUnary(not, stack); }
	 else if(strcmp(token,"add")== 0)      { applyBinary(add, stack); }
	 else if(strcmp(token,"sub")== 0)      { applyBinary(sub, stack); }
	 else if(strcmp(token,"mul")== 0)      { applyBinary(mul, stack); }
	 else if(strcmp(token,"div")== 0)      { applyBinary(divide, stack); }
	 else if(strcmp(token,"rem")== 0)      { applyBinary(rem, stack); }
	 else if(strcmp(token,"neg")== 0)      { applyUnary(negate, stack); }
	 else if(strcmp(token,"pop")== 0)      { popStack(stack); }
	 else if(strcmp(token,"exc")== 0)      { excStack(stack); }
	 else if(strcmp(token,"bind")== 0)      { applyBindings(stack, bindings); }
	 else if(strcmp(token,"if")== 0)      { applyIf(stack); }
	 else if(strcmp(token,"load")== 0)      { 
		 while (1){
			struct Type * x = pop(stack);
			char y[80];
			int i = 0;
			if(x->type != STRING){
				push(stack,x);
				push(stack,makeError());
				break;
			}
			 FILE *opened_file;
			 opened_file =fopen(x->value.string,"r");
			 if(opened_file == NULL){
				push(stack,x);
				push(stack,makeError());
				break;
			 }
			 int c;
			 while((c = fgetc(opened_file)) != EOF){
				 if(c == ' ' || c == '\n'){
					char z[i+1];
					strncpy(z,y,i+1);
					eval(z,stack,bindings);
					memset(&y,0,sizeof(y));
					i=0;
				 }else{
					y[i] = c;
					i = i + 1;
				 }
			 }
			 push(stack,makeBoolean(TRUE));
			 break;
		 }
	 }
	 else if(strcmp(token,"quit")== 0)     { exit(0); }
	 else if(strspn(token, "0123456789-") == strlen(token)) { push(stack,makeNumber(atoi(token))); }
	 else if(isLetter(token) == TRUE) { push(stack,makeName(token, bindings)); }
	 else { push(stack,makeError()); }
	 return stack;
}

/*
 * Main repl function
 * updated to incorporate new features
 */

void repl(){

	 char * buffer;
	 char * token;
	 char storeBuffer[BUFFER_LENGTH];
	 size_t buffLength = 0;
	 struct Node *stack;
	 stack = makeEmptyList(); // make a stack to hold tokens
	 map *bindings;  // make map to hold bindings
	 bindings = (map *) (malloc(sizeof(map)));
	 bindings->start = NULL;


	 while(1){
	 printf("repl>");
	 getline(&buffer,&buffLength,stdin);
	 strcpy(storeBuffer,buffer);
	 storeBuffer[strlen(storeBuffer)-1] = '\0';
		
	 /* get the first token */
	 token = strtok(storeBuffer, " \n\t");

	 /* walk through other tokens */
	 while( token != NULL ) {
		stack = eval(token,stack, bindings);
		token = strtok(NULL, " \n\t");
	 }		 
	 printStack(stack);
	 }
}

/*   Main Method   */

int main(){
  repl();
  return 0;
}
